node-uniffi/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+//! Native library providing Rust to mobile language bindings for the Lumina node.
+//!
+//! This crate uses Mozillas UniFFI to generate Swift and Kotlin bindings for the Lumina node,
+//! allowing it to be used from iOS and Android applications.
+#![cfg(not(target_arch = "wasm32"))]
+
+mod error;
+mod types;
+
+use celestia_types::ExtendedHeader;
+use error::{LuminaError, Result};
+use lumina_node::{
+    blockstore::RedbBlockstore, events::EventSubscriber, node::PeerTrackerInfo, store::RedbStore,
+    Node,
+};
+use std::str::FromStr;
+use tendermint::hash::Hash;
+use tokio::sync::{Mutex, RwLock};
+use types::{NetworkInfo, NodeConfig, NodeEvent, PeerId, SyncingInfo};
+use uniffi::Object;
+
+uniffi::setup_scaffolding!();
+
+lumina_node::uniffi_reexport_scaffolding!();
+
+/// The main Lumina node that manages the connection to the Celestia network.
+#[derive(Object)]
+pub struct LuminaNode {
+    node: RwLock<Option<Node<RedbBlockstore, RedbStore>>>,
+    events_subscriber: Mutex<Option<EventSubscriber>>,
+    config: NodeConfig,
+}
+
+#[uniffi::export(async_runtime = "tokio")]
+impl LuminaNode {
+    /// Sets a new connection to the Lumina node for the specified network.
+    #[uniffi::constructor]
+    pub fn new(config: NodeConfig) -> Result<Self> {
+        Ok(Self {
+            node: RwLock::new(None),
+            events_subscriber: Mutex::new(None),
+            config,
+        })
+    }
+
+    /// Starts the node and connects to the network.
+    pub async fn start(&self) -> Result<bool> {
+        let mut node_lock = self.node.write().await;
+        if node_lock.is_some() {
+            return Err(LuminaError::NetworkError {
+                msg: "Node is already running".to_string(),
+            });
```

Comment:
I think it would less boilerplate to do one of the following in all errors that have a description message:

1. Use tuple instead. For example `LuminaError::NetworkError(String)`

2. If you want to keep the `msg` field then write some helpers. For example:

```rust
impl LuminaError {
    pub(crate) fn network_error(msg: &str) -> LuminaError {
        LuminaError::NetworkError { msg: msg.to_owned() }
    }
}
```

And then use it as:

```rust
return Err(LuminaError::network_error("Node is already running"));
```

Also, another thing, since `LuminaError` contains the `Error` in its name I prefer one of the two:

1. `LuminaError::Network`
2. `LuminaError::NetworkFailure`